### DISCUSSION TITLE
Weird behavior when form contains button which open <ModalForm />

### DISCUSSION THREAD

**Discussion by Steffi3rd** [react-hook-form/react-hook-form#3704] [Q&A] [answered]
Hello,

I have a form where I have 2 buttons : one to submit the current form and another to open a `<ModalForm />`.

In this `<ModalForm />`, I have another form.

The weird behavior is when I submit form inside `<ModalForm />`, it submits the parent form whereas when I check the tree in the browser, the modal is not inside the parent form but at the end of page just before `</body>`

```
const { handleSubmit } = useForm(...);

return (
  <form onSubmit={handleSubmit(...)}>
      ...
      <OpenModalForm />
      <button type="submit">Submit</button>
  </form>
)
```

```
const OpenModalForm = () => (
  const [modalOpen, setOpenModal] = React.useState(false);

  return (
    <>
      <button onClick={() => setOpenModal(true)}>
        Open Modal Form
      </button>
      <ModalForm
        open={modalOpen}
        onClose={() => setOpenModal(false)}
      />
    </>
  );
)
```

```
const ModalForm = () => (
  const { handleSubmit } = useForm(...);

  return (
    <form onSubmit={handleSubmit(...)}>
        ...
        <button type="submit">Submit</button>
     </form>
  );
)
```

Is there a solution for that?

**Comment by bluebill1049** [comment]
This is correct you should not nest your form inside of each other.

You can have several forms on a page but they should not be nested.

> 4.10.3 The form element
> 
> Content model:
> 
> Flow content, but with no form element descendants.

**Reply by Steffi3rd** [reply]
Yup, I know that. But the thing is, it seems that my `<ModalForm />` is not really nested because it appears at the end of page just before `</body>` (so outside of the parent `<form>`)

What do you think?

**Reply by bluebill1049** [reply]
not sure on that then, can you reproduce that in a simple cSB?

**Reply by adamayres** [reply]
@bluebill1049 I ran into this same issue and have created a codesandox with the reproduction here: https://codesandbox.io/s/react-hook-form-nested-9qtck

**Some things to note**

* The form in the Modal is added to the DOM as a child of the `<body>` element, so it is not a nested form with regards to the DOM. The modal, and the form it contains, are however child components of the containing form, so as far as the React tree is concerned, they are nested. The modal and it's form are added to the `<body>` by way of using a React Portal, here is a [reference](https://github.com/react-bootstrap/react-overlays/blob/master/src/Modal.tsx#L360) to the code where that is done.
* I believe the issue here is that the Synthetic "submit" event is bubbling up the React Tree (see [here](https://github.com/facebook/react/issues/11387) and [here](https://github.com/facebook/react/issues/19637)) causing the modal in the form to submit and the parent component that contains a form to submit.
* Attempting to call `event.stopPropagation()` inside of the `onValid` callback of `handleSubmit` of `useForm` does not prevent the form in the containing component from submitting. However, if you call `event.stopPropagation()` before calling `handleSubmit` in the modal form then it does prevent the form in the containing component from being submitted (see workaround below). I believe this may happen because the `onValid` callback in `handleSubmit` is called [async](https://github.com/react-hook-form/react-hook-form/blob/master/src/useForm.ts#L1142) - after the event has already bubbled.
* This same issue occurs whether or not you use `FormProvider` or directly use `useForm`. This same issue has been raised before [here](https://github.com/react-hook-form/react-hook-form/discussions/3675) where the recommendation was to move the form in the modal/dialog outside of the containing form. I believe the reason this fixes the issue is because the form is no longer nested in the React tree and thus the synthetic event does not bubble up to the containing component with the form. This is a reasonable workaround however is not viable given the way I have my code organized.

**Workaround**

In the nested form in the modal, change this:

```js
<Form onSubmit={handleSubmit(onValid, onInvalid)}>
```

To this:

```js
<Form 
  onSubmit={(event) => {
    event.stopPropagation();
    handleSubmit(onValid, onInvalid)(event)
  }}
>
```

**Potential Fix in library**

In [`handleSubmit`](https://github.com/react-hook-form/react-hook-form/blob/0a0f2bf959aff7065dee18137d6fc9f26a082644/src/useForm.ts#L1076-L1081) you might consider calling `event.stopPropagation` OR expose a prop that conditionally calls it. Thoughts?

**Reply by jonasgroendahl** [reply]
genius, thanks!

**Comment by bluebill1049** [comment]
thanks @adamayres if you want to post another comment and we can mark https://github.com/react-hook-form/react-hook-form/discussions/3704#discussioncomment-265857 as answer

**Comment by adamayres** [comment, marked as answer]
Reposting as a top level comment, originally posted [here](https://github.com/react-hook-form/react-hook-form/discussions/3704#discussioncomment-265857).

I ran into this same issue and have created a codesandox with the reproduction here: https://codesandbox.io/s/react-hook-form-nested-9qtck

**Some things to note**

* The form in the Modal is added to the DOM as a child of the `<body>` element, so it is not a nested form with regards to the DOM. The modal, and the form it contains, are however child components of the containing form, so as far as the React tree is concerned, they are nested. The modal and it's form are added to the `<body>` by way of using a React Portal, here is a [reference](https://github.com/react-bootstrap/react-overlays/blob/master/src/Modal.tsx#L360) to the code where that is done.
* I believe the issue here is that the Synthetic "submit" event is bubbling up the React Tree (see [here](https://github.com/facebook/react/issues/11387) and [here](https://github.com/facebook/react/issues/19637)) causing the modal in the form to submit and the parent component that contains a form to submit.
* Attempting to call `event.stopPropagation()` inside of the `onValid` callback of `handleSubmit` of `useForm` does not prevent the form in the containing component from submitting. However, if you call `event.stopPropagation()` before calling `handleSubmit` in the modal form then it does prevent the form in the containing component from being submitted (see workaround below). I believe this may happen because the `onValid` callback in `handleSubmit` is called [async](https://github.com/react-hook-form/react-hook-form/blob/master/src/useForm.ts#L1142) - after the event has already bubbled.
* This same issue occurs whether or not you use `FormProvider` or directly use `useForm`. This same issue has been raised before [here](https://github.com/react-hook-form/react-hook-form/discussions/3675) where the recommendation was to move the form in the modal/dialog outside of the containing form. I believe the reason this fixes the issue is because the form is no longer nested in the React tree and thus the synthetic event does not bubble up to the containing component with the form. This is a reasonable workaround however is not viable given the way I have my code organized.

**Workaround**

In the nested form in the modal, change this:

```js
<Form onSubmit={handleSubmit(onValid, onInvalid)}>
```

To this:

```js
<Form 
  onSubmit={(event) => {
    event.stopPropagation();
    handleSubmit(onValid, onInvalid)(event)
  }}
>
```

**Potential Fix in library**

In [`handleSubmit`](https://github.com/react-hook-form/react-hook-form/blob/0a0f2bf959aff7065dee18137d6fc9f26a082644/src/useForm.ts#L1076-L1081) you might consider calling `event.stopPropagation` OR expose a prop that conditionally calls it. Thoughts?

**Reply by bluebill1049** [reply]
> In [handleSubmit](https://github.com/react-hook-form/react-hook-form/blob/0a0f2bf959aff7065dee18137d6fc9f26a082644/src/useForm.ts#L1076-L1081) you might consider calling event.stopPropagation OR expose a prop that conditionally calls it. Thoughts?

Yes, we can consider that as a config.

**Reply by hornta** [reply]
Have you come to any conclusion regarding making this an option?
